# `NYC Parking Group Case Study -- Apache Spark`

##### The purpose of this case study is to conduct an exploratory data analysis that will help us understand the data that NYC Police Department has collected for parking tickets. For the scope of this analysis, we will analyse the parking tickets over the year 2017. 

In [ ]:
__author__ = 'Sri Harsha Ravi'

## Creating a PySpark Session

In [1]:
# class pyspark.sql.SparkSession, The entry point to programming Spark with the Dataset and DataFrame API.
#A SparkSession can be used create DataFrame and perform several actions on it.

from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("Group Case Study - NYC Parking").getOrCreate()

In [49]:
spark

## Reading Parking Ticket data into a PySpark DataFrame

In [2]:
# Importing data from a csv file located in the HDFS into a PySpark dataframe
parkingDF = spark.read.format("csv")\
  .option("header", "true")\
  .option("inferSchema", "true")\
  .load("/common_folder/nyc_parking/Parking_Violations_Issued_-_Fiscal_Year_2017.csv")
# cache dataframe for better performance
parkingDF.cache()

DataFrame[Summons Number: bigint, Plate ID: string, Registration State: string, Issue Date: timestamp, Violation Code: int, Vehicle Body Type: string, Vehicle Make: string, Violation Precinct: int, Issuer Precinct: int, Violation Time: string]

In [3]:
# printSchema returns schema in tree format
parkingDF.printSchema()

# Displaying the first 5 rows of the dataframe
parkingDF.show(5)

root
 |-- Summons Number: long (nullable = true)
 |-- Plate ID: string (nullable = true)
 |-- Registration State: string (nullable = true)
 |-- Issue Date: timestamp (nullable = true)
 |-- Violation Code: integer (nullable = true)
 |-- Vehicle Body Type: string (nullable = true)
 |-- Vehicle Make: string (nullable = true)
 |-- Violation Precinct: integer (nullable = true)
 |-- Issuer Precinct: integer (nullable = true)
 |-- Violation Time: string (nullable = true)

+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+
|Summons Number|Plate ID|Registration State|         Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Violation Precinct|Issuer Precinct|Violation Time|
+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+
|    5092469481| GZH7067|                NY|201

## Examine the data

### 1. Find the total number of tickets for the year.

##### Filtering data to select only where tickets related to 2017

*For the scope of this analysis, we will analyse the parking tickets over the year 2017. So we will filter the dataframe for tickets belonging to only 2017.*

In [4]:
#we will also register this as a temporary view so that we can query it with SQL and show off basic transformations in SQL
parkingDF.registerTempTable("parking_2017")

# Creating a new column called Issue_Yr by taking year part from 'Issue Date'
parking2017DF = spark.sql("SELECT *, substr(`Issue Date`, 1,4) as Issue_Yr FROM parking_2017 where substr(`Issue Date`, 1,4) = '2017'")
parking2017DF.show(5)

+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+
|Summons Number|Plate ID|Registration State|         Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Violation Precinct|Issuer Precinct|Violation Time|Issue_Yr|
+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+
|    8478629828| 66623ME|                NY|2017-06-14 00:00:00|            47|             REFG|       MITSU|                14|             14|         1120A|    2017|
|    5096917368| FZD8593|                NY|2017-06-13 00:00:00|             7|             SUBN|       ME/BE|                 0|              0|         0852P|    2017|
|    1407740258| 2513JMG|                NY|2017-01-11 00:00:00|            78|             DELV|       FRUEH|               106|            106|     

In [5]:
# Printing total no of tickets before and after filtering for Issue Yr = 2017
from pyspark.sql.functions import count
from pyspark.sql.functions import countDistinct
parkingDF.select(countDistinct("Summons Number").alias("total_tickets_before_filter")).show()
parking2017DF.select(countDistinct("Summons Number").alias("total_tickets_in_2017")).show()

+---------------------------+
|total_tickets_before_filter|
+---------------------------+
|                   10803028|
+---------------------------+

+---------------------+
|total_tickets_in_2017|
+---------------------+
|              5431918|
+---------------------+



* Total no of tickets in the dataframe = `10803028`
* Total no of tickets in the dataframe for the year 2017 = `5431918`

### 2. Find out the number of unique states from where the cars that got parking tickets came.

##### Displaying the top States with cars that got highest parking tickets

In [53]:
# Calculating count of tickets after grouping the dataframe by "Registration State" and sorting in descending order
from pyspark.sql.functions import expr, col, column
from pyspark.sql.functions import col, asc, desc
parking2017DF.groupBy("Registration State").agg(expr("count(`Summons Number`)")\
                                                .alias("total_tickets")).sort(desc("total_tickets")).show(5)

Py4JJavaError: An error occurred while calling o540.showString.
: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
Exchange hashpartitioning(Registration State#811, 200)
+- *(1) HashAggregate(keys=[Registration State#811], functions=[partial_count(Summons Number#10L)], output=[Registration State#811, count#3482L])
   +- *(1) Project [Summons Number#10L, CASE WHEN (cast(Registration State#12 as int) = 99) THEN NY ELSE Registration State#12 END AS Registration State#811]
      +- *(1) Filter (((((isnotnull(Issue Date#13) && isnotnull(Violation Time#19)) && (substring(cast(Issue Date#13 as string), 1, 4) = 2017)) && NOT (Violation Time#19 = nan)) && NOT (Violation Time#19 = Nan)) && NOT (CASE WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 0) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 400)) THEN Dawn WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 400) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 800)) THEN Early Mornig WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 800) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 1200)) THEN Morning Rush WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 1200) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 1600)) THEN Afternoon WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 1600) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 2000)) THEN Office Time WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 2000) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) <= 2400)) THEN Late Night ELSE unknown END = unknown))
         +- InMemoryTableScan [Issue Date#13, Registration State#12, Summons Number#10L, Violation Time#19], [isnotnull(Issue Date#13), isnotnull(Violation Time#19), (substring(cast(Issue Date#13 as string), 1, 4) = 2017), NOT (Violation Time#19 = nan), NOT (Violation Time#19 = Nan), NOT (CASE WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 0) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 400)) THEN Dawn WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 400) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 800)) THEN Early Mornig WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 800) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 1200)) THEN Morning Rush WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 1200) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 1600)) THEN Afternoon WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 1600) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) < 2000)) THEN Office Time WHEN ((cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) >= 2000) && (cast(CASE WHEN (substring(Violation Time#19, -1, 1) = A) THEN substring(Violation Time#19, 1, 4) WHEN ((substring(Violation Time#19, -1, 1) = P) && (substring(Violation Time#19, 1, 2) = 12)) THEN concat(00, substring(Violation Time#19, 3, 2)) ELSE concat(cast(cast((cast(substring(Violation Time#19, 1, 2) as double) + 12.0) as int) as string), substring(Violation Time#19, 3, 2)) END as int) <= 2400)) THEN Late Night ELSE unknown END = unknown)]
               +- InMemoryRelation [Summons Number#10L, Plate ID#11, Registration State#12, Issue Date#13, Violation Code#14, Vehicle Body Type#15, Vehicle Make#16, Violation Precinct#17, Issuer Precinct#18, Violation Time#19], StorageLevel(disk, memory, deserialized, 1 replicas)
                     +- *(1) FileScan csv [Summons Number#10L,Plate ID#11,Registration State#12,Issue Date#13,Violation Code#14,Vehicle Body Type#15,Vehicle Make#16,Violation Precinct#17,Issuer Precinct#18,Violation Time#19] Batched: false, Format: CSV, Location: InMemoryFileIndex[hdfs://nameservice1/common_folder/nyc_parking/Parking_Violations_Issued_-_Fisca..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<Summons Number:bigint,Plate ID:string,Registration State:string,Issue Date:timestamp,Viola...

	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.doExecute(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.InputAdapter.inputRDDs(WholeStageCodegenExec.scala:374)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.inputRDDs(HashAggregateExec.scala:151)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:610)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.TakeOrderedAndProjectExec.executeCollect(limit.scala:136)
	at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collectFromPlan(Dataset.scala:3384)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2545)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2545)
	at org.apache.spark.sql.Dataset$$anonfun$53.apply(Dataset.scala:3365)
	at org.apache.spark.sql.execution.SQLExecution$$anonfun$withNewExecutionId$1.apply(SQLExecution.scala:78)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:125)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:73)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3364)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:2545)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:2759)
	at org.apache.spark.sql.Dataset.getRows(Dataset.scala:255)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:292)
	at sun.reflect.GeneratedMethodAccessor70.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$.needToCopyObjectsBeforeShuffle(ShuffleExchangeExec.scala:163)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$.prepareShuffleDependency(ShuffleExchangeExec.scala:303)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.prepareShuffleDependency(ShuffleExchangeExec.scala:91)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:128)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 39 more


*New York has got the most no of cars with parking tickets.*

##### Displaying the distinct values of such States and total no of States

In [9]:
spark.sql("SELECT distinct `Registration State` FROM parking_2017").sort(asc("Registration State")).show()
parking2017DF.select(countDistinct("Registration State")).show()

+------------------+
|Registration State|
+------------------+
|                99|
|                AB|
|                AK|
|                AL|
|                AR|
|                AZ|
|                BC|
|                CA|
|                CO|
|                CT|
|                DC|
|                DE|
|                DP|
|                FL|
|                FO|
|                GA|
|                GV|
|                HI|
|                IA|
|                ID|
+------------------+
only showing top 20 rows

+----------------------------------+
|count(DISTINCT Registration State)|
+----------------------------------+
|                                65|
+----------------------------------+



*Now there is a numeric entry '99' in the column, which should be corrected and as per business requirement replaced with the state having the maximum entries - in this case which is `NY` *

In [10]:
from pyspark.sql.functions import when

parking2017DF = parking2017DF.withColumn("Registration State", \
                                       when(parking2017DF["Registration State"] == 99, 'NY').\
                                       otherwise(parking2017DF["Registration State"]))

parking2017DF.select(countDistinct("Registration State")).show()

+----------------------------------+
|count(DISTINCT Registration State)|
+----------------------------------+
|                                64|
+----------------------------------+



*After replacing '99' by `NY`, total no of unique states = `64`*

In [11]:
# update the SQL view to record the changes done to the dataframe
parking2017DF.registerTempTable("parking_2017")

##### Deleting null values

In [12]:
dropNullDF = parking2017DF.na.drop()

In [13]:
#from pyspark.sql.functions import countDistinct
parking2017DF.select(countDistinct("Summons Number")).show()
dropNullDF.select(countDistinct("Summons Number")).show()

+------------------------------+
|count(DISTINCT Summons Number)|
+------------------------------+
|                       5431918|
+------------------------------+

+------------------------------+
|count(DISTINCT Summons Number)|
+------------------------------+
|                       5431918|
+------------------------------+



*There are no null values in the dataframe.*

## Aggregation tasks

### 1. How often does each violation code occur? Display the frequency of the top five violation codes.

##### Grouping the data based on Violation Code and displaying the top 5 occurences

In [14]:
from pyspark.sql.functions import sum, count, avg, expr
parking2017DF.groupBy("Violation Code").agg(expr("count(`Summons Number`)").\
                                           alias("total_violations")).sort(desc("total_violations")).show(5)

+--------------+----------------+
|Violation Code|total_violations|
+--------------+----------------+
|            21|          768087|
|            36|          662765|
|            38|          542079|
|            14|          476664|
|            20|          319646|
+--------------+----------------+
only showing top 5 rows



### 2. How often does each 'vehicle body type' get a parking ticket? How about the 'vehicle make'? (Hint: Find the top 5 for both.)

##### Grouping the data based on Vehicle Body Type and Vehicle Make and displaying the top 5 results for both

In [15]:
parking2017DF.groupBy("Vehicle Body Type").agg(expr("count(`Summons Number`)").\
                                           alias("total_violations")).sort(desc("total_violations")).show(5)
                                           
parking2017DF.groupBy("Vehicle Make").agg(expr("count(`Summons Number`)").\
                                           alias("total_violations")).sort(desc("total_violations")).show(5)                                           

+-----------------+----------------+
|Vehicle Body Type|total_violations|
+-----------------+----------------+
|             SUBN|         1883954|
|             4DSD|         1547312|
|              VAN|          724029|
|             DELV|          358984|
|              SDN|          194197|
+-----------------+----------------+
only showing top 5 rows

+------------+----------------+
|Vehicle Make|total_violations|
+------------+----------------+
|        FORD|          636844|
|       TOYOT|          605291|
|       HONDA|          538884|
|       NISSA|          462017|
|       CHEVR|          356032|
+------------+----------------+
only showing top 5 rows



### 3. A precinct is a police station that has a certain zone of the city under its command. Find the (5 highest) frequencies of tickets for each of the following:
    1.'Violation Precinct' (This is the precinct of the zone where the violation occurred). Using this, can you draw any 
        insights for parking violations in any specific areas of the city?
    2.'Issuer Precinct' (This is the precinct that issued the ticket.) 
        Here, you would have noticed that the dataframe has the'Violating Precinct' or 'Issuing Precinct' as '0'. These are 
        erroneous entries. Hence, you need to provide the records for five correct precincts.
        (Hint: Print the top six entries after sorting.)

##### Grouping the data based on 'Violation Precinct' & 'Issuer Precinct' and displaying the top 6 occurences

In [16]:
parking2017DF.groupBy("Violation Precinct").agg(expr("count(`Summons Number`)").\
                                           alias("total_violations")).sort(desc("total_violations")).show(6)
                                           
parking2017DF.groupBy("Issuer Precinct").agg(expr("count(`Summons Number`)").\
                                           alias("total_violations")).sort(desc("total_violations")).show(6)

+------------------+----------------+
|Violation Precinct|total_violations|
+------------------+----------------+
|                 0|          925596|
|                19|          274445|
|                14|          203553|
|                 1|          174702|
|                18|          169131|
|               114|          147444|
+------------------+----------------+
only showing top 6 rows

+---------------+----------------+
|Issuer Precinct|total_violations|
+---------------+----------------+
|              0|         1078406|
|             19|          266961|
|             14|          200495|
|              1|          168740|
|             18|          162994|
|            114|          144054|
+---------------+----------------+
only showing top 6 rows



### 4. Find the violation code frequencies for three precincts that have issued the most number of tickets. Do these precinct zones have an exceptionally high frequency of certain violation codes? Are these codes common across precincts? 
(Hint: In the SQL view, use the 'where' attribute to filter among three precincts.)

##### Displaying the three precincts that have issued the most number of tickets.

In [17]:
spark.sql("SELECT `Issuer Precinct`, COUNT(*) AS total_violation \
FROM parking_2017 \
WHERE `Issuer Precinct` != 0 \
group by `Issuer Precinct` \
order by total_violation desc limit 3").show()

+---------------+---------------+
|Issuer Precinct|total_violation|
+---------------+---------------+
|             19|         266961|
|             14|         200495|
|              1|         168740|
+---------------+---------------+



##### Displaying the 5 most commonly occuring violation codes from the three precincts that have issued the most number of tickets.

In [18]:
spark.sql("select `Issuer Precinct`, `Violation Code`, total_violation, RANK  from \
          (select *, row_number() over (PARTITION by `Issuer Precinct` order by total_violation desc) as RANK from \
            (SELECT b.`Violation Code`, b.`Issuer Precinct`, COUNT(*) AS total_violation \
            FROM (SELECT `Issuer Precinct`, COUNT(*) AS total_violation \
            FROM parking_2017 \
            WHERE `Issuer Precinct` != 0 \
            group by `Issuer Precinct` \
            order by total_violation desc limit 3) as a \
            inner join parking_2017 as b on a.`Issuer Precinct` = b.`Issuer Precinct` \
            group by b.`Violation Code`,b.`Issuer Precinct`) as c) where RANK < 6").show()

+---------------+--------------+---------------+----+
|Issuer Precinct|Violation Code|total_violation|RANK|
+---------------+--------------+---------------+----+
|              1|            14|          38354|   1|
|              1|            16|          19081|   2|
|              1|            20|          15408|   3|
|              1|            46|          12745|   4|
|              1|            38|           8535|   5|
|             19|            46|          48445|   1|
|             19|            38|          36386|   2|
|             19|            37|          36056|   3|
|             19|            14|          29797|   4|
|             19|            21|          28415|   5|
|             14|            14|          45036|   1|
|             14|            69|          30464|   2|
|             14|            31|          22555|   3|
|             14|            47|          18364|   4|
|             14|            42|          10027|   5|
+---------------+-----------

##### The most commonly occuring violation code for different Precincts are -
 - Precinct 1 -> 14
 - Precinct 14 -> 14 
 - Precinct 19 -> 46
 
*The violation code - 14 seems to be a common factor in the top 5 most occuring violation codes for the 3 precincts* 

### 5. Find out the properties of parking violations across different times of the day:
 - Find a way to deal with missing values, if any.
(Hint: Check for the null values using 'isNull' under the SQL. Also, to remove the null values, check the 'dropna' command in the API documentation.)

 - The Violation Time field is specified in a strange format. Find a way to make this a time attribute that you can use to divide into groups.

 - Divide 24 hours into six equal discrete bins of time. Choose the intervals as you see fit. For each of these groups, find the three most commonly occurring violations.
(Hint: Use the CASE-WHEN in SQL view to segregate into bins. To find the most commonly occurring violations, you can use an approach similar to the one mentioned in the hint for question 4.)

 - Now, try another direction. For the three most commonly occurring violation codes, find the most common time of the day (in terms of the bins from the previous part).

##### Displaying no of occurences of the following in the dataframe --

 - NULL
 - BLANK
 - Nan/nan

In [19]:
dropNullDF1 = parking2017DF.filter(col("Violation Time").isNull())
dropNullDF2 = parking2017DF.where("`Violation Time` == ''")
dropNullDF3 = parking2017DF.where((col("Violation Time") == "nan") | (col("Violation Time") == "Nan"))
dropNullDF1.select(countDistinct("Violation Time"), count("Violation Time")).show()
dropNullDF2.select(countDistinct("Violation Time"), count("Violation Time")).show()
dropNullDF3.select(countDistinct("Violation Time"), count("Violation Time")).show()

+------------------------------+---------------------+
|count(DISTINCT Violation Time)|count(Violation Time)|
+------------------------------+---------------------+
|                             0|                    0|
+------------------------------+---------------------+

+------------------------------+---------------------+
|count(DISTINCT Violation Time)|count(Violation Time)|
+------------------------------+---------------------+
|                             0|                    0|
+------------------------------+---------------------+

+------------------------------+---------------------+
|count(DISTINCT Violation Time)|count(Violation Time)|
+------------------------------+---------------------+
|                             1|                   16|
+------------------------------+---------------------+



*There seems to be 16 records in the dataframe that have the value 'Nan' or 'nan' in the column 'Violation Time'. We will be dropping these records from the dataframe.* 

In [20]:
parking2017DF = parking2017DF.filter((col("Violation Time") != "nan") & (col("Violation Time") != "Nan"))

parking2017DF.select(countDistinct("Violation Time"), count("Violation Time")).show()

+------------------------------+---------------------+
|count(DISTINCT Violation Time)|count(Violation Time)|
+------------------------------+---------------------+
|                          1625|              5431902|
+------------------------------+---------------------+



*After dropping 'Nan/nan' records from the dataframe, there are 5431902 records left.* 

In [21]:
# update the SQL view to record the changes done to the dataframe
parking2017DF.registerTempTable("parking_2017")

##### The Violation Time field is specified in a strange format (Eg. 1120A). We are transforming it as per the below logic in a new column Violation Time Fixed:

 - If the `Violation Time` value ends with 'A', then we just take the 1st 4 digits and ignore the alphabet 'A'. Eg. 1120A -> 1120.
 - If the `Violation Time` value ends with 'P', and the first 2 digits are '12' then we concatenate '00' with the 3rd and 4th digit and ignore the alphabet 'P'. Eg. 1220P -> 0020.
 - For all other types of occurences, ie. if the `Violation Time` value ends with 'P', and the first 2 digits aren't '12' then we add '12' to the first 2 digits and concatenate it with the 3rd and 4th digit and ignore the alphabet 'P'. Eg. 0852P -> 2052.

In [24]:
parking2017DF = spark.sql("SELECT *, \
                        CASE WHEN substr(`Violation Time`, -1,1) == 'A' \
                            THEN substr(`Violation Time`, 1,4) \
                            WHEN substr(`Violation Time`, -1,1) == 'P' AND substr(`Violation Time`, 1,2) == '12'\
                            THEN CONCAT('00',substr(`Violation Time`, 3,2)) \
                            ELSE CONCAT(CAST((substr(`Violation Time`, 1,2) + 12) as int),substr(`Violation Time`, 3,2)) \
                            END AS `Violation Time Fixed` \
FROM parking_2017")
parking2017DF.show(5)

+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+--------------------+
|Summons Number|Plate ID|Registration State|         Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Violation Precinct|Issuer Precinct|Violation Time|Issue_Yr|Violation Time Fixed|
+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+--------------------+
|    8478629828| 66623ME|                NY|2017-06-14 00:00:00|            47|             REFG|       MITSU|                14|             14|         1120A|    2017|                1120|
|    5096917368| FZD8593|                NY|2017-06-13 00:00:00|             7|             SUBN|       ME/BE|                 0|              0|         0852P|    2017|                2052|
|    1407740258| 2513JMG|                NY|2

In [25]:
# update the SQL view to record the changes done to the dataframe
parking2017DF.registerTempTable("parking_2017")

##### Dividing the 24 hours into six equal discrete bins of time based on the new column Violation Time Fixed:
 - Between 0 and 400 --> Dawn
 - Between 400 and 800 --> Early Mornig
 - Between 800 and 1200 --> Morning Rush
 - Between 1200 and 1600 --> Afternoon
 - Between 1600 and 2000 --> Office Time
 - Between 2000 and 2400 --> Late Night
 - For everything else --> unknown

In [26]:
parking2017DF = spark.sql("SELECT *, \
                        CASE WHEN `Violation Time Fixed` >= 0 and `Violation Time Fixed` < 400\
                            THEN 'Dawn' \
                            WHEN `Violation Time Fixed` >= 400 and `Violation Time Fixed` < 800\
                            THEN 'Early Mornig' \
                            WHEN `Violation Time Fixed` >= 800 and `Violation Time Fixed` < 1200\
                            THEN 'Morning Rush' \
                            WHEN `Violation Time Fixed` >= 1200 and `Violation Time Fixed` < 1600\
                            THEN 'Afternoon' \
                            WHEN `Violation Time Fixed` >= 1600 and `Violation Time Fixed` < 2000\
                            THEN 'Office Time' \
                            WHEN `Violation Time Fixed` >= 2000 and `Violation Time Fixed` <= 2400\
                            THEN 'Late Night' \
                            ELSE 'unknown' \
                            END AS `Violation Time Span` \
FROM parking_2017")
parking2017DF.show(5)

+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+--------------------+-------------------+
|Summons Number|Plate ID|Registration State|         Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Violation Precinct|Issuer Precinct|Violation Time|Issue_Yr|Violation Time Fixed|Violation Time Span|
+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+--------------------+-------------------+
|    8478629828| 66623ME|                NY|2017-06-14 00:00:00|            47|             REFG|       MITSU|                14|             14|         1120A|    2017|                1120|       Morning Rush|
|    5096917368| FZD8593|                NY|2017-06-13 00:00:00|             7|             SUBN|       ME/BE|                 0|              0|         08

In [28]:
# update the SQL view to record the changes done to the dataframe
parking2017DF.registerTempTable("parking_2017")

##### Checking the count of records where Violation Time Span = unknown

In [34]:
spark.sql("select `Violation Time`,`Violation Time Fixed`, `Violation Time Span`  \
from parking_2017 where `Violation Time Span` == 'unknown'").show(5)

spark.sql("select count(*) from parking_2017 where `Violation Time Span` == 'unknown'").show(5)

+--------------+--------------------+-------------------+
|Violation Time|Violation Time Fixed|Violation Time Span|
+--------------+--------------------+-------------------+
|         6815P|                8015|            unknown|
|         110+A|                110+|            unknown|
|         093+A|                093+|            unknown|
|         8715P|                9915|            unknown|
|         5402P|                6602|            unknown|
+--------------+--------------------+-------------------+
only showing top 5 rows



##### There are 73 records where Violation Time Span = unknown. These are essentially faulty records and need to be removed. For eg, Violation Time values for some of these records are `6815P`, `110+A`, `093+A`, etc. Thus these records could not be binned into Violation Time Span.

In [36]:
parking2017DF = parking2017DF.filter(col("`Violation Time Span`") != "unknown")
parking2017DF.show(5)
parking2017DF.select(countDistinct("Violation Time Span"), count("Violation Time Span")).show()

+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+--------------------+-------------------+
|Summons Number|Plate ID|Registration State|         Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Violation Precinct|Issuer Precinct|Violation Time|Issue_Yr|Violation Time Fixed|Violation Time Span|
+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+--------------------+-------------------+
|    8478629828| 66623ME|                NY|2017-06-14 00:00:00|            47|             REFG|       MITSU|                14|             14|         1120A|    2017|                1120|       Morning Rush|
|    5096917368| FZD8593|                NY|2017-06-13 00:00:00|             7|             SUBN|       ME/BE|                 0|              0|         08

##### After removal of faulty records, the no of rows in the dataframe  = 5431829

In [37]:
# update the SQL view to record the changes done to the dataframe
parking2017DF.registerTempTable("parking_2017")

##### Displaying the 3 most commonly occuring violation codes for each of the 6 different Violation Time Spans

In [38]:
spark.sql("select `Violation Time Span`, `Violation Code`, no_of_violations from \
         (select *, row_number() over (PARTITION by `Violation Time Span` order by no_of_violations desc) as num from \
         (select `Violation Time Span`, `Violation Code`, count(*) as no_of_violations \
         from parking_2017 group by `Violation Time Span`, `Violation Code`) as a) \
         where num < 4").show()

+-------------------+--------------+----------------+
|Violation Time Span|Violation Code|no_of_violations|
+-------------------+--------------+----------------+
|               Dawn|            21|          107065|
|               Dawn|            36|          101991|
|               Dawn|            38|           56204|
|        Office Time|            38|          102855|
|        Office Time|            14|           75902|
|        Office Time|            37|           70345|
|          Afternoon|            38|          184829|
|          Afternoon|            36|          184293|
|          Afternoon|            37|          130692|
|       Early Mornig|            14|           74113|
|       Early Mornig|            40|           60652|
|       Early Mornig|            21|           57894|
|         Late Night|             7|           26293|
|         Late Night|            40|           22336|
|         Late Night|            14|           21045|
|       Morning Rush|       

##### Displaying the most common Violation Time Span of the day for the three most commonly occurring violation codes

In [39]:
spark.sql("select`Violation Code`,`Violation Time Span`,total_violation from \
          (select *, row_number() over (PARTITION by `Violation Code` order by total_violation desc) as num from \
            (SELECT b.`Violation Code`, b.`Violation Time Span`, COUNT(*) AS total_violation \
            FROM (SELECT `Violation Code`, COUNT(*) AS total_violation \
            FROM parking_2017 \
            group by `Violation Code` \
            order by total_violation desc limit 3) as a \
            inner join parking_2017 as b on a.`Violation Code` = b.`Violation Code` \
            group by b.`Violation Code`,b.`Violation Time Span`) as c) where num < 2").show()

+--------------+-------------------+---------------+
|Violation Code|Violation Time Span|total_violation|
+--------------+-------------------+---------------+
|            38|          Afternoon|         184829|
|            21|       Morning Rush|         598062|
|            36|       Morning Rush|         348165|
+--------------+-------------------+---------------+



### 6. Let’s try and find some seasonality in this data:

 - First, divide the year into a certain number of seasons, and find the frequencies of tickets for each season. (Hint: Use Issue Date to segregate into seasons.)

 - Then, find the three most common violations for each of these seasons. (Hint: You can use an approach similar to the one mentioned in the hint for question 4.)

In [40]:
# update the SQL view to record the changes done to the dataframe
parking2017DF.registerTempTable("parking_2017")

##### Dividing the year into following number of seasons based on the column Issue Date into a new column  Issue Season:
 - Month between 3 and 5 --> Spring
 - Month between 6 and 8 --> Summer
 - Month between 9 and 11 --> Fall
 - Else --> Winter

In [42]:
parking2017DF = spark.sql("SELECT *, \
CASE WHEN substr(`Issue Date`,6,2) between 3 and 5 \
    THEN 'Spring' \
    WHEN substr(`Issue Date`,6,2) between 6 and 8 \
    THEN 'Summer' \
    WHEN substr(`Issue Date`,6,2) between 9 and 11 \
    THEN 'Fall' \
    ELSE 'Winter' \
    END AS `Issue Season` \
                        FROM parking_2017")
parking2017DF.show(5)

+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+--------------------+-------------------+------------+
|Summons Number|Plate ID|Registration State|         Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Violation Precinct|Issuer Precinct|Violation Time|Issue_Yr|Violation Time Fixed|Violation Time Span|Issue Season|
+--------------+--------+------------------+-------------------+--------------+-----------------+------------+------------------+---------------+--------------+--------+--------------------+-------------------+------------+
|    8478629828| 66623ME|                NY|2017-06-14 00:00:00|            47|             REFG|       MITSU|                14|             14|         1120A|    2017|                1120|       Morning Rush|      Summer|
|    5096917368| FZD8593|                NY|2017-06-13 00:00:00|             7|             SUBN|       

In [44]:
# update the SQL view to record the changes done to the dataframe
parking2017DF.registerTempTable("parking_2017")

##### Displaying the 3 most commonly occuring violation codes for each of the 4 seasons

In [45]:
spark.sql("select `Issue Season`, `Violation Code`, no_of_violations from \
         (select *, row_number() over (PARTITION by `Issue Season` order by no_of_violations desc) as num from \
         (select `Issue Season`, `Violation Code`, count(*) as no_of_violations \
         from parking_2017 group by `Issue Season`, `Violation Code`) as a) \
         where num < 4").show()

+------------+--------------+----------------+
|Issue Season|Violation Code|no_of_violations|
+------------+--------------+----------------+
|      Spring|            21|          402401|
|      Spring|            36|          344834|
|      Spring|            38|          271167|
|      Summer|            21|          127345|
|      Summer|            36|           96663|
|      Summer|            38|           83518|
|        Fall|            46|             231|
|        Fall|            21|             128|
|        Fall|            40|             116|
|      Winter|            21|          238180|
|      Winter|            36|          221268|
|      Winter|            38|          187385|
+------------+--------------+----------------+



### 7. The fines collected from all the instances of parking violation constitute a source of revenue for the NYC Police Department. Let’s take an example of estimating this for the three most commonly occurring codes:
 - Find the total occurrences of the three most common violation codes.
 - Then, visit the website: http://www1.nyc.gov/site/finance/vehicles/services-violation-codes.page . It lists the fines associated with different violation codes. They’re divided into two categories: one for the highest-density locations in the city and the other for the rest of the city. For the sake of simplicity, take the average of the two.
 - Using this information, find the total amount collected for the three violation codes with the maximum tickets. State the code that has the highest total collection.
 - What can you intuitively infer from these findings?

##### Finding the total occurrences of the three most common violation codes.

In [46]:
top3violationcodes = spark.sql("select `Violation Code`, count(*) as no_of_violations \
         from parking_2017 group by `Violation Code` order by no_of_violations desc limit 3")
top3violationcodes.show()

+--------------+----------------+
|Violation Code|no_of_violations|
+--------------+----------------+
|            21|          768054|
|            36|          662765|
|            38|          542078|
+--------------+----------------+



In [47]:
# register DataFrame as temp table
top3violationcodes.registerTempTable("violation_codes_df") 

##### Listing the fines associated with the 3 different violation codes : one for the highest-density locations in the city and the other for the rest of the city.

In [48]:
top3violationcodes = spark.sql("SELECT *, \
                                CASE WHEN `Violation Code` == 21 \
                                THEN  65 \
                                WHEN `Violation Code` == 36 \
                                THEN 50 \
                                ELSE 65 \
                                END AS `High_Density_Loc_Fine`, \
                                CASE WHEN `Violation Code` == 21 \
                                THEN  45 \
                                WHEN `Violation Code` == 36 \
                                THEN 50 \
                                ELSE 35 \
                                END AS `Rest_of_City_Fine` \
                                FROM violation_codes_df")
top3violationcodes.show(5)           

+--------------+----------------+---------------------+-----------------+
|Violation Code|no_of_violations|High_Density_Loc_Fine|Rest_of_City_Fine|
+--------------+----------------+---------------------+-----------------+
|            21|          768054|                   65|               45|
|            36|          662765|                   50|               50|
|            38|          542078|                   65|               35|
+--------------+----------------+---------------------+-----------------+



In [49]:
# update the SQL view to record the changes done to the dataframe
top3violationcodes.registerTempTable("violation_codes_df")

##### Taking the average of the 2 categories of fines for each of the 3 different violation codes.

In [50]:
top3violationcodes = spark.sql("SELECT *, \
                                 cast((`High_Density_Loc_Fine` + `Rest_of_City_Fine`)/2 as int) AS `Average Fine` \
                                FROM violation_codes_df")
top3violationcodes.show(5)           

+--------------+----------------+---------------------+-----------------+------------+
|Violation Code|no_of_violations|High_Density_Loc_Fine|Rest_of_City_Fine|Average Fine|
+--------------+----------------+---------------------+-----------------+------------+
|            21|          768054|                   65|               45|          55|
|            36|          662765|                   50|               50|          50|
|            38|          542078|                   65|               35|          50|
+--------------+----------------+---------------------+-----------------+------------+



In [49]:
# update the SQL view to record the changes done to the dataframe
top3violationcodes.registerTempTable("violation_codes_df")

##### Finding the total amount collected for the three violation codes with the maximum tickets by multiplying the average fine per violation to the total occurences of each violation

In [52]:
spark.sql("SELECT *, `Average Fine` * no_of_violations as `Total Fine` FROM violation_codes_df").show()

+--------------+----------------+---------------------+-----------------+------------+----------+
|Violation Code|no_of_violations|High_Density_Loc_Fine|Rest_of_City_Fine|Average Fine|Total Fine|
+--------------+----------------+---------------------+-----------------+------------+----------+
|            21|          768054|                   65|               45|          55|  42242970|
|            36|          662765|                   50|               50|          50|  33138250|
|            38|          542078|                   65|               35|          50|  27103900|
+--------------+----------------+---------------------+-----------------+------------+----------+



*Violation Code - 21 has the highest total collection in terms of fine*

### Stopping the SparkContext Object

In [53]:
spark.stop()